### ISSUE TITLE
Add NEFTune to improve fine-tuning? 

### ISSUE THREAD

**Issue by nkasmanoff** [Lightning-AI/lit-gpt#642]
Came across this thread / paper / repo on Twitter yesterday and it seems like a promising approach for getting a little more out of your fine-tuning. Haven't tested for myself just yet, but can report those results unless someone else already has. 

I started to implement a version of this on my fork of lit-gpt, but wondering if this is of broader interest? Can make this a PR if so. 

links to the work:
https://twitter.com/neeljain1717/status/1712559136188272982
https://arxiv.org/abs/2310.05914
https://github.com/neelsjain/NEFTune

my implementation (not yet tested but I think you get the gist)
https://github.com/nkasmanoff/llm-finetuning/commit/93527f2523edf081e5e427a4336415a2d45109c4

**Comment by sanyalsunny111** [comment]
Hey @nkasmanoff just curious did you get time to test this?

**Comment by nkasmanoff** [comment]
Hey @sanyalsunny111, TLDR I did for a very simple example, and think it should work for LoRA. If you want to try it, all you'd need to do is specify `neftune_alpha` in `finetune/lora.py` (currently default to `None`)
I updated the link above & corresponding PR. 

The simple example: 
I fine-tuned Llama2 7B for 50 steps with a LoRA adapter only on the LM head. Using the logs I confirmed that the logic in the PR only adds noise to the embeddings during training, and the result was slightly worse train loss for these extremely early on parts of fine-tuning. Admittedly there's a lot more you'd want to confirm like what performance gains this has during actual training, but I don't have the compute or time budget to check that all out right now.

**Comment by sanyalsunny111** [comment]
Great!! Thanks

**Comment by carmocca** [comment]
There's no plan to add this implementation for now, but I'm glad that you were able to easily implement it with Lit-GPT
